### PR TITLE
 Circulation: Policy configuration

### PR DESCRIPTION
**Pull request by BadrAly**
* NEW: add circulation policy objects

Co-Authored-by: Aly Badr <aly.badr@rero.ch>
Co-Authored-by: Gianni Pante <gianni.pante@rero.ch>

## To test

- get the PR branch
- `./scripts/bootstrap` 
- `./scripts/setup` 
- run the tests

- Display the list of available circulation policies (cipo), add, delete and edit

### CODE REVIEW

File: data/circulation_policies.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[
+  {
+    "$schema": "https://ils.test.rero.ch/schema/circ_policies/circ_policy-v0.0.1.json",
+    "name": "standard",
+    "description": "default standard circulation policy",
```
Comment (iGormilhit) [inline data/circulation_policies.json]: Start the description with a capital and end it with a dot.

File: data/circulation_policies.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    "allow_checkout": true,
+    "checkout_duration": 30,
+    "allow_requests": true,
+    "number_renewals": 3,
+    "renewal_duration": 30
+  },
+  {
+    "$schema": "https://ils.test.rero.ch/schema/circ_policies/circ_policy-v0.0.1.json",
+    "name": "short",
+    "description": "short circulation policy",
```
Comment (iGormilhit) [inline data/circulation_policies.json]: Start the description with a capital and end it with a dot.

File: data/circulation_policies.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    "allow_checkout": true,
+    "checkout_duration": 10,
+    "allow_requests": true,
+    "number_renewals": 1,
+    "renewal_duration": 10
+  },
+  {
+    "$schema": "https://ils.test.rero.ch/schema/circ_policies/circ_policy-v0.0.1.json",
+    "name": "on-site",
+    "description": "on-site circulation policy",
```
Comment (iGormilhit) [inline data/circulation_policies.json]: Start the description with a capital and end it with a dot.

File: rero_ils/modules/circ_policies/form_options/circ_policies/circ_policy-v0.0.1.json [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    "title": "Circulation Policy",
+    "items": [
+      {
+        "key": "pid",
+        "readonly": true
+      },
+      {
+        "key": "name",
+        "validationMessage": {
+          "alreadyTakenMessage": "The circulation policy name is already taken"
```
Comment (iGormilhit) [inline rero_ils/modules/circ_policies/form_options/circ_policies/circ_policy-v0.0.1.json]: Missing dot at the end of the message.

File: rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json [hunk 5]
```diff
@@ -0,0 +1,71 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "title": "CircPolicy",
```
Comment (iGormilhit) [inline rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json]: Why not `Circulation policy`?

File: rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  "description": "JSON schema for circ_policies.",
```
Comment (iGormilhit) [inline rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json]: Should be `JSON schema for circulation policies.`

File: rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    "$schema",
+    "pid",
+    "name",
+    "description",
+    "organisation_pid"
+  ],
+  "properties": {
+    "$schema": {
+      "title": "Schema",
+      "description": "Schema to validate circ_policies records against.",
```
Comment (iGormilhit) [inline rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json]: Change `circ_policies` to `circulation policies`.

File: rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      "type": "string",
+      "minLength": 9
+    },
+    "pid": {
+      "title": "CircPolicy ID",
```
Comment (iGormilhit) [inline rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json]: Change `CircPolicy` to `Circulation policy`. And check all the JSON schema with this rule.

File: rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+      "type": "string",
+      "minLength": 1
+    },
+    "name": {
+      "title": "CircPolicy name",
+      "description": "The name of the circulation policy",
```
Comment (iGormilhit) [inline rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json]: Missing dot at the end of the message. And check all the description messages.

File: rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      "minLength": 2
+    },
+    "description": {
+      "title": "CircPolicy description",
+      "description": "The description of the circulation policy",
+      "type": "string"
+    },
+    "organisation_pid": {
+      "title": "Organisation ID",
+      "description": "The organisation_pid pid of this circulation policy",
```
Comment (iGormilhit) [inline rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json]: `The PID of the organisation in which this circulation policy applies.` Or something like it.

File: rero_ils/modules/circ_policies/minters.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+"""Identifier minters."""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+from functools import partial
+
+from ..minters import id_minter
+from .providers import CircPolicyProvider
+
+circ_policies_id_minter = partial(
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/minters.py]: singular on policies ?

File: rero_ils/modules/circ_policies/models.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+from __future__ import absolute_import
+
+from invenio_db import db
+from invenio_pidstore.models import RecordIdentifier
+
+
+class CircPolicyIdentifier(RecordIdentifier):
+    """Sequence generator for CircPolicy identifiers."""
+
+    __tablename__ = 'circ_policies_id'
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/models.py]: singular on policies ?

File: rero_ils/modules/circ_policies/templates/rero_ils/detailed_view_circ_policies.html [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        {% include 'rero_ils/_editor_button_actions.html' %}
+    {% endwith %}
+
+</div>
+{%- endblock %}
+
+{%- block javascript %}
+{{ super() }}
+{% assets "rero_ils_tooltips_js" %}<script src="{{ ASSET_URL }}"></script>{% endassets %}
+{%- endblock javascript %}
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/templates/rero_ils/detailed_view_circ_policies.html]: Indent is 2 spaces (file)

File: rero_ils/modules/circ_policies/views.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+
+@blueprint.route('/circ_policies/name/validate/<name>', methods=["GET"])
+@record_edit_permission.require()
+def name_validate(name):
+    """Circ policy name Validate."""
+    response = {
+        'name': None,
+        'organisation_pid': None
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/views.py]: Remove this (unused)

File: rero_ils/config.py [hunk 15]
```diff
@@ -490,8 +491,30 @@ def _(x):
         default_media_type='application/json',
         max_result_window=10000,
         search_factory_imp='rero_ils.query:and_search_factory'
+    ),
+    cipo=dict(
+        pid_type='cipo',
+        pid_minter='circ_policies_id',
```
Comment (Garfield-fr) [inline rero_ils/config.py]: singular on policies ?

File: rero_ils/config.py [hunk 16]
```diff
@@ -490,8 +491,30 @@ def _(x):
...
+        pid_fetcher='circ_policies_id',
```
Comment (Garfield-fr) [inline rero_ils/config.py]: singular on policies ?

File: rero_ils/modules/circ_policies/api.py [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+    class Meta:
+        """Search only on patrons index."""
+
+        index = 'circ_policies'
+
+
+class CircPolicy(IlsRecord):
+    """CircPolicy class."""
+
+    minter = circ_policies_id_minter
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/api.py]: singular on policies ?

File: rero_ils/modules/circ_policies/api.py [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+    fetcher = circ_policies_id_fetcher
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/api.py]: singular on policies ?

File: rero_ils/modules/circ_policies/fetchers.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+
+from __future__ import absolute_import, print_function
+
+from functools import partial
+
+from ..fetchers import id_fetcher
+from .providers import CircPolicyProvider
+
+circ_policies_id_fetcher = partial(
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/fetchers.py]: singular on policies ?

File: rero_ils/modules/circ_policies/templates/rero_ils/circ_policy_editor.html [hunk 20]
```diff
@@ -0,0 +1,9 @@
+{%- extends config.RERO_ILS_EDITOR_TEMPLATE %}
+
+{%- block schema_editor %}
+<div class="container">
+    <div class="row">
+        <schema-form ng-controller="CircPolicyValidator"></schema-form>
+    </div>
+</div>
+{%- endblock schema_editor %}
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/templates/rero_ils/circ_policy_editor.html]: indent is 2 spaces (file)

File: setup.py [hunk 21]
```diff
@@ -177,7 +179,8 @@ def run(self):
             'patron_id = rero_ils.modules.patrons.minters:patron_id_minter',
             'patron_type_id = \
                 rero_ils.modules.patrons_types.minters:patron_type_id_minter',
-            'mef_person_id = rero_ils.modules.mef.minters:mef_person_id_minter'
+            'mef_person_id = rero_ils.modules.mef.minters:mef_person_id_minter',
+            'circ_policies_id = rero_ils.modules.circ_policies.minters:circ_policies_id_minter',
```
Comment (Garfield-fr) [inline setup.py]: Singular on policies ?

File: setup.py [hunk 22]
```diff
@@ -197,7 +200,8 @@ def run(self):
             'patron_type_id = \
                 rero_ils.modules.patrons_types.fetchers:patron_type_id_fetcher',
             'mef_person_id = \
-                rero_ils.modules.mef.fetchers:mef_person_id_fetcher'
+                rero_ils.modules.mef.fetchers:mef_person_id_fetcher',
+            'circ_policies_id = rero_ils.modules.circ_policies.fetchers:circ_policies_id_fetcher',
```
Comment (Garfield-fr) [inline setup.py]: Singular on policies ?

File: tests/modules/circ_policies/test_circ_policies_fetchers.py [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Fetchers module tests."""
+
+from __future__ import absolute_import, print_function
+
+from uuid import uuid4
+
+from rero_ils.modules.circ_policies.fetchers import \
+    circ_policies_id_fetcher as fetcher
```
Comment (Garfield-fr) [inline tests/modules/circ_policies/test_circ_policies_fetchers.py]: Singular on policies ?

File: tests/modules/circ_policies/test_circ_policies_fetchers.py [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+from rero_ils.modules.circ_policies.minters import \
+    circ_policies_id_minter as minter
```
Comment (Garfield-fr) [inline tests/modules/circ_policies/test_circ_policies_fetchers.py]: Singular on policies ?

File: tests/modules/circ_policies/test_circ_policies_minters.py [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+
+from rero_ils.modules.circ_policies import minters
+
+
+def test_circ_policies_id_minter(db):
+    """Test policies_id minter."""
+    data = {}
+    # first record
+    rec_uuid = uuid4()
+    pid1 = minters.circ_policies_id_minter(rec_uuid, data)
```
Comment (Garfield-fr) [inline tests/modules/circ_policies/test_circ_policies_minters.py]: Singular on policies ?

File: tests/modules/circ_policies/test_circ_policies_minters.py [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+    assert pid1
+    assert data['pid'] == pid1.pid_value
+    assert data['pid'] == '1'
+    assert pid1.object_type == 'rec'
+    assert pid1.object_uuid == rec_uuid
+
+    # second record
+    data = {}
+    rec_uuid = uuid4()
+    pid2 = minters.circ_policies_id_minter(rec_uuid, data)
```
Comment (Garfield-fr) [inline tests/modules/circ_policies/test_circ_policies_minters.py]: Singular on policies ?

File: rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+    "allow_requests": {
+      "title": "Allow patron requests flag",
+      "description": "Patron requests are allowed or not",
+      "type": "boolean",
+      "default": true
+    },
+    "number_renewals": {
+      "title": "Maximum number of renewals",
+      "description": "Maximum number of renewals allowed",
+      "type": "integer"
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json]: default value ?

File: rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+    },
+    "renewal_duration": {
+      "title": "Renewal duration in days",
+      "description": "The duration of the renewal in days",
+      "type": "integer"
+    }
```
Comment (Garfield-fr) [inline rero_ils/modules/circ_policies/jsonschemas/circ_policies/circ_policy-v0.0.1.json]: default value ?

**Pull request merged**